### ISSUE TITLE
Add rug plot option

### ISSUE THREAD

**Issue by bdeonovic** [dcjones/Gadfly.jl#315]
Will there be support for a geom_rug as ggplot 2 has? 

http://docs.ggplot2.org/0.9.2.1/geom_rug.html

**Comment by bdeonovic** [comment]
I was not too impressed by what the plot looked like. 

```julia
using Gadfly
x = randn(100)
p = plot(x=x, Geom.density, Guide.xrug)
draw(PNG("test.png", 6inch, 3inch), p)
```
![test 2](https://cloud.githubusercontent.com/assets/805168/10955534/14f9af72-831a-11e5-9ded-65e576052a65.png)

something similar in R looks like

```R
x <- rnorm(100)
plot(density(x))
rug(x)
```

![rplot01](https://cloud.githubusercontent.com/assets/805168/10955551/3f11137c-831a-11e5-92ac-c3ad4a9df22f.png)

In Gadfly all of the rug ticks seem to blend together and you cannot see where they are clustering.

**Comment by dcjones** [comment]
Yeah, I didn't fully think this through. The issue is that the ticks are plotting the points where the density curve is evaluated, no the original data. I don't know of the work around, I think the solution is to implement rug as a geometry rather than a guide.

**Comment by bdeonovic** [comment]
Should we reopen for visibility of the issue?

**Comment by bdeonovic** [comment]
Is this something difficult to change?

**Comment by bdeonovic** [comment]
Bump, anyone interested in changed rug from a guide to a geometry?

**Comment by tlnagy** [comment]
I never understood why rug was a guide, not a geometry. @Mattriks do you have time to look into this? I might have time this weekend, but I'm spread kinda thin right now.

**Comment by Mattriks** [comment]
This is wip, the following code can be turned into `Geom.rug` in the future (i.e. this is a step towards `Geom.rug`). Because a marginal rug is really a type of `SegmentGeometry`, we can (for the moment) combine `Stat_rug` with `Geom.segment`. The examples below will work now, the idea here is for people to provide feedback on the design, particularly the arguments. For example, I could add an orientation argument, e.g. `Stat_rug(orientation=[ :horizontal, :vertical]`, such as in `Geom.bar`. What do people think?

```julia
 type RugStatistic <: Gadfly.StatisticElement
    min::Vector{AbstractFloat}
    length::AbstractFloat
end

function Gadfly.input_aesthetics(stat::RugStatistic)
    return [:x]
end

function Gadfly.output_aesthetics(stat::RugStatistic)
    return [:x, :y, :xend, :yend]
end

function Gadfly.default_scales(stat::RugStatistic)
    return [Gadfly.Scale.x_continuous()]
end


function Gadfly.Stat.apply_statistic(stat::RugStatistic,
                         scales::Dict{Symbol, Gadfly.ScaleElement},
                         coord::Gadfly.CoordinateElement,
                         aes::Gadfly.Aesthetics)
    nsides = length(stat.min)
    n = length(aes.x)
    aes.x = aes.x
    aes.y = fill(stat.min[1], n)
    aes.xend = aes.x
    aes.yend = fill(stat.min[1]+stat.length, n)
    
    if nsides>1
        aes.x = [aes.x; aes.x]
        append!(aes.y, fill(stat.min[2], n))
        aes.xend = [aes.xend; aes.xend]
        append!(aes.yend, fill(stat.min[2]-stat.length, n))
        aes.color = [aes.color; aes.color]    
    end
end
```

## Examples
```julia
theme(x) = Theme(default_color=parse(Colors.Colorant,x))
Stat_rug = RugStatistic
```

### Example 1
The first argument to `Stat_rug()` is a vector

```julia
srand(123)
p1 = plot(x=randn(100),
    layer(Geom.density), 
    layer(Stat_rug([0.0], 0.025), Geom.segment, theme("grey")),
)
```
![i315ex1](https://cloud.githubusercontent.com/assets/18226881/18697513/3dbc0ebc-8005-11e6-8b9f-a8cdcd2ce442.png)

### Example 2
Use a vector of length two for [bottom, top] rugs

```julia
srand(123)
id = rand(["a","b"], 100)
x = randn(100)
x[id.=="a"] += 2.0

p2 = plot(x=x, color=id,
    layer(Geom.density), 
    layer(Stat_rug([0.0, 0.5], 0.025), Geom.segment)
#    layer(Stat_rug([NaN, 0.5], 0.025), Geom.segment) # top rug only
)
```
![i315ex2](https://cloud.githubusercontent.com/assets/18226881/18697536/6d5b5c2c-8005-11e6-951c-209fc36cb7b7.png)

**Comment by tlnagy** [comment]
Thanks @Mattriks, that looks very cool. I don't know if having a vector as one of the parameters is the best, it seems more complicated than necessary. I like the overall aesthetic though.

**Comment by Mattriks** [comment]
Easy - add an outer function to the first code block above:
```julia
RugStatistic{T<:AbstractFloat}(min::T,length::T) = RugStatistic([min],length)
```
So this is now possible:
```julia
layer(Stat_rug(0.6, -0.025), Geom.segment, theme("grey")),
```
The idea of using numbers is that the user has a lot of control over where the rug is placed - in the above layer, the rug is placed much higher (=0.6) than the top axis (=0.4) in Example 1.

**Comment by tlnagy** [comment]
Would it be possible to add some sane defaults, like if someone just does `Stat.rug` then it just figures out the best place to put it? Otherwise, I think this would be a great update.

**Comment by Mattriks** [comment]
Look at my first code block above (`Gadfly.default_scales`) and try `dump(Scale.x_continuous())`. `minvalue` and `maxvalue` currently default to `nothing` (most Geometries don't need it), unless you manually provide values through `Scale.x_` etc.  As a Gadfly user, I wouldn't want to do the latter if I could more simply do:
```julia
layer(Geom.rug([0.0, 0.6], 0.025))
```
which is the next step in `Geom.rug` development!

**Comment by tlnagy** [comment]
Ugh. I see. That's annoying that it doesn't pass through real values.

**Comment by bdeonovic** [comment]
This looks like a good solution to me, is this being merged in?

**Comment by Mattriks** [comment]
Not yet, there are some things to do here. Unfortunately, no time at the moment. But as a workaround, the above code will work for simple plots.

**Comment by tlnagy** [comment]
Everything has cleared up with the CI now. I would love to add this if someone has the time to submit a PR.

**Comment by Mattriks** [comment]
I think it might be possible also to implement `Geom.rug` using relative and not absolute plot coordinates, and this should be tested first before submitting a PR.

**Comment by bdeonovic** [comment]
Any update on this matter?

**Comment by bjarthur** [comment]
@Mattriks there's been a request to tag a new release by @tkelman .  can you get to this soon or should we just go ahead and do so now?

**Comment by Mattriks** [comment]
No, not soon. Yes, tag a new release!

**Comment by bdeonovic** [comment]
Any update?
